CMakeLists.txt

```diff
@@ -0,0 +1,563 @@
+    "from torch.utils.cpp_extension import COMMON_NVCC_FLAGS; print(';'.join(COMMON_NVCC_FLAGS))"
+    "Failed to determine torch nvcc compiler flags")
+
+  if (CUDA_VERSION VERSION_GREATER_EQUAL 11.8)
+    list(APPEND VLLM_GPU_FLAGS "-DENABLE_FP8_E5M2")
+  endif()
+
+  if(NVCC_THREADS)
+    list(APPEND VLLM_GPU_FLAGS "--threads=${NVCC_THREADS}")
+  endif()
+
+  set(VLLM_PUNICA_GPU_FLAGS ${VLLM_GPU_FLAGS})
+
+  #
+  # Copy flags+update for punica
+  #
+  list(REMOVE_ITEM VLLM_PUNICA_GPU_FLAGS
+    "-D__CUDA_NO_HALF_OPERATORS__"
+    "-D__CUDA_NO_HALF_CONVERSIONS__"
+    "-D__CUDA_NO_BFLOAT16_CONVERSIONS__"
+    "-D__CUDA_NO_HALF2_OPERATORS__")
+
+  message(DEBUG "nvcc: ${VLLM_GPU_FLAGS}")
+  message(DEBUG "punica nvcc: ${VLLM_PUNICA_GPU_FLAGS}")
+
+  #
+  # Setup/process CUDA arch flags.
+  #
+  # The torch cmake setup hardcodes the detected architecture flags in
+  # `CMAKE_CUDA_FLAGS`.  Since `CMAKE_CUDA_FLAGS` is a "global" variable, it
+  # can't modified on a per-target basis, e.g. for the `punica` extension.
+  # So, all the `-gencode` flags need to be extracted and removed from
+  # `CMAKE_CUDA_FLAGS` for processing so they can be passed by another method.
+  # Since it's not possible to use `target_compiler_options` for adding target
+  # specific `-gencode` arguments, the target's `CUDA_ARCHITECTURES` property
+  # must be used instead.  This requires repackaging the architecture flags
+  # into a format that cmake expects for `CUDA_ARCHITECTURES`.
+  #
+  # This is a bit fragile in that it depends on torch using `-gencode` as opposed
+  # to one of the other nvcc options to specify architectures.
+  #
+  # Note: torch uses the `TORCH_CUDA_ARCH_LIST` environment variable to override
+  # detected architectures.
+  #
+  message(DEBUG "initial CMAKE_CUDA_FLAGS: ${CMAKE_CUDA_FLAGS}")
+
+  # Extract all `-gencode` flags from `CMAKE_CUDA_FLAGS`
+  string(REGEX MATCHALL "-gencode arch=[^ ]+" VLLM_CUDA_ARCH_FLAGS
+    ${CMAKE_CUDA_FLAGS})
+
+  # Remove all `-gencode` flags from `CMAKE_CUDA_FLAGS` since they will be modified
+  # and passed back via the `CUDA_ARCHITECTURES` property.
+  string(REGEX REPLACE "-gencode arch=[^ ]+ *" "" CMAKE_CUDA_FLAGS
+    ${CMAKE_CUDA_FLAGS})
+
+  # If this error is triggered, it might mean that torch has changed how it sets
+  # up nvcc architecture code generation flags.
+  if (NOT VLLM_CUDA_ARCH_FLAGS)
+    message(FATAL_ERROR
+      "Could not find any architecture related code generation flags in "
+      "CMAKE_CUDA_FLAGS. (${CMAKE_CUDA_FLAGS})")
+  endif()
+
+  message(DEBUG "arch flags: ${VLLM_CUDA_ARCH_FLAGS}")
+
+  # Macro for converting a `gencode` version number to a cmake version number.
+  macro(string_to_ver OUT_VER IN_STR)
+    string(REGEX REPLACE "\([0-9]+\)\([0-9]\)" "\\1.\\2" ${OUT_VER} ${IN_STR})
+  endmacro()
+
+  # Initialize the architecture lists to empty.
+  set(VLLM_GPU_ARCHES)
+  set(VLLM_PUNICA_GPU_ARCHES)
+
+  # Process each `gencode` flag.
+  foreach(ARCH ${VLLM_CUDA_ARCH_FLAGS})
```

Comment:
thank you so much for supporting the existing workflow of `VLLM_CUDA_ARCH_FLAGS`. greatly appreciated!